seckill-query-service/src/main/java/io/servicecomb/poc/demo/seckill/SeckillEventLoader.java

```diff
@@ -0,0 +1,119 @@
+
+  private SpringBasedPromotionEventRepository promotionEventRepository;
+  private PromotionRepository promotionRepository;
+
+  private final Map<String, List<Coupon>> customerCoupons = new HashMap<>();
+  private final List<Promotion> activePromotions = new ArrayList<>();
+
+  SeckillEventLoader(SpringBasedPromotionEventRepository promotionEventRepository,
+      PromotionRepository promotionRepository) {
+    this.promotionEventRepository = promotionEventRepository;
+    this.promotionRepository = promotionRepository;
+  }
+
+  void reloadEventsScheduler() {
+    ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
+    executor.scheduleWithFixedDelay(new Runnable() {
+                                      int currentPromotionEventIndex = 0;
+
+                                      @Override
+                                      public void run() {
+                                        List<PromotionEvent> promotionEvents = promotionEventRepository
+                                            .findByIdGreaterThan(currentPromotionEventIndex);
+
+                                        updateActivePromotions(promotionEvents);
+
+                                        for (PromotionEvent promotionEvent : promotionEvents) {
+                                          reloadActivePromotions(promotionEvent);
+                                          reloadSuccessCoupons(promotionEvent);
+
+                                          currentPromotionEventIndex = promotionEvent.getId();
+                                        }
+                                      }
+                                    },
+        0,
+        1000,
+        TimeUnit.MILLISECONDS);
+  }
+
+  public List<Coupon> getCustomerCoupons(String customerId) {
+    return customerCoupons.get(customerId);
+  }
+
+  public List<Promotion> getActivePromotions() {
+    return activePromotions;
+  }
+
+  private void reloadSuccessCoupons(PromotionEvent promotionEvent) {
+    customerCoupons.computeIfAbsent((String)promotionEvent.getCustomerId(), customerId -> new ArrayList<>()).add(
+        new Coupon<>(promotionEvent.getId(),
+            promotionEvent.getPromotionId(),
+            promotionEvent.getTime(),
+            promotionEvent.getDiscount(),
+            promotionEvent.getCustomerId())
+    );
+  }
+
+  private void updateActivePromotions(List<PromotionEvent> promotionEvents){
+    List<PromotionEvent> finishPromotionEvents = promotionEvents.stream().filter(
+        promotionEvent -> promotionEvent.getType().equals(PromotionEventType.Finish)).collect(Collectors.toList());
+
+    for (int i = 0; i < activePromotions.size(); i++) {
+      Promotion activePromotion = activePromotions.get(i);
+
+      boolean isMatch = finishPromotionEvents.stream().anyMatch(
+          finishEvent -> finishEvent.getPromotionId().equals(activePromotion.getPromotionId()));
+      if (isMatch) {
+        activePromotions.remove(activePromotion);
```

Comment:
activePromotions is both being modified and iterated. it may throw concurrent modification exception. please use iterator instead.

Reply:
fixed